src/install/hosted_git_info.zig

```diff
@@ -0,0 +1,1807 @@
+                // We can't really guess either -- there's no :// which comes before @
+                return .{ .url = npa_str, .protocol = .unknown };
+            }
+        } else {
+            // Something like user@host which is also a valid URL. Since no :, that means that the
+            // URL is as good as it gets. No need to slice.
+            return .{ .url = npa_str, .protocol = .{ .well_formed = .git_plus_ssh } };
+        }
+    }
+
+    // The next thing we can try is to search for the double slash and treat this protocol as a
+    // custom one.
+    //
+    // NOTE(markovejnovic): I also think this is wrong in parse-url.js.
+    // They:
+    // 1. Test the protocol against known protocols (which is fine)
+    // 2. Then, if not found, they go through that hoop of checking for @ and : guessing if it is a
+    //    git+ssh URL or not
+    // 3. And finally, they search for ://.
+    //
+    // The last two steps feel like they should happen in reverse order:
+    //
+    // If I have a foobar://user:host@path URL (and foobar is not given as a known protocol), their
+    // implementation will not report this as a foobar protocol, but rather as
+    // git+ssh://foobar://user:host@path which, I think, is wrong.
+    //
+    // I even tested it: https://tinyurl.com/5y4e6zrw
+    //
+    // Our goal is to be bug-for-bug compatible, at least for now, so this is how I re-implemented
+    // it.
+    const maybe_dup_slash_idx = bun.strings.indexOf(npa_str, "//");
+    if (maybe_dup_slash_idx) |dup_slash_idx| {
+        if (dup_slash_idx == first_colon_idx + 1) {
+            return .{
+                .url = bun.strings.substring(npa_str, dup_slash_idx + 2, null),
+                .protocol = .{ .custom = npa_str[0..dup_slash_idx] },
+            };
+        }
+    }
+
+    // Well, otherwise we have to split the original URL into two pieces,
+    // right at the colon.
+    if (first_colon_idx != -1) {
+        return .{
+            .url = bun.strings.substring(npa_str, @intCast(first_colon_idx + 1), null),
+            .protocol = .{ .custom = npa_str[0..@intCast(first_colon_idx + 1)] },
+        };
+    }
+
+    // Well we couldn't figure out anything.
+    return .{ .url = npa_str, .protocol = .unknown };
+}
+
+/// Attempt to correct an scp-style URL into a proper URL, parsable with jsc.URL. Potentially
+/// mutates the original input.
+///
+/// This function assumes that the input is an scp-style URL.
+fn correctUrlMut(url_proto_pair: UrlProtocolPair) UrlProtocolPair {
+    const at_idx: isize = if (bun.strings.lastIndexBeforeChar(url_proto_pair.url, '@', '#')) |idx|
+        @intCast(idx)
+    else
+        -1;
+
+    const col_idx: isize = if (bun.strings.lastIndexBeforeChar(url_proto_pair.url, ':', '#')) |idx|
+        @intCast(idx)
+    else
+        -1;
+
+    if (col_idx > at_idx) {
+        url_proto_pair.url[@intCast(col_idx)] = '/';
```

Comment:
is there any way we can remove this mutation, or allocate at the one place this function is called if it needs to add a `'/'`? if we can change this then all `[]u8` in this file can become `[]const u8`, which will simplify adding it to the rest of bun install. for example, in dependnecy.zig we won't need to clone the `dependnecy` string before each `HostedGitInfo.fromUrl`